# Gini Imputity Calculation

## Gini impurity on performance split

In [1]:
print(1 - ((0.33 ** 2) + (0.67 ** 2)))

0.4421999999999999


In [2]:
print(1 - ((0.57 ** 2) + (0.43 ** 2)))

0.4902000000000001


In [3]:
print(((14 / 20) * 0.49) +((6 / 20) * 0.44))

0.475


0.475 Gini imppuruty in Performance split

## Gini impurity based on class split

20 students > 10 in 9th standard  -> 2 play, 8 don't play. 10 in 10th standard -> 8 play, 2 don't play


In [5]:
print(1 - (((2 / 10) ** 2) + ((8 / 10 ) ** 2)))

0.31999999999999984


In [6]:
print(1 - (((8 / 10) ** 2) + ((2 / 10 ) ** 2)))

0.31999999999999984


In [7]:
print(((10 / 20) * 0.32) + ((10 / 20) *0.32))

0.32


0.32 Gini impurity on class split

In [8]:
print(1 / 7)

0.14285714285714285


# Chi Square for Best Split

Chi square = √(〖(Actual −Expected)〗^ 2 / Expected) 

In [13]:
from math import sqrt

## Chi Sqaure on Perfomance Split

In [18]:
sqrt(1/7)


0.3779644730092272

In [19]:
sqrt(1/3)

0.5773502691896257

Chi square of Class = 0.38 + 0.38 + 0.58 + 0.58 = 1.16 + 0.76 = 1.92

## Chi Square on Class Split

In [20]:
sqrt(9 / 5)

1.3416407864998738

In [21]:
sqrt(9 / 5)

1.3416407864998738

Chi square of Class = 1.34 + 1.34 + 1.34 + 1.34 = 5.36

### High Chi Square for Class Split Hence best split

# Information Gain 

Information Gain = 1 - Entropy

Where,

Entropy = − p1〖log〗_2 p1 − p2 〖log〗_2 p2 −… − pn〖log〗_2 pn



## Entropy Based on Perfomance Split

In [22]:
from math import log2

In [24]:
- (0.5 * log2(0.5)) - (0.5 * log2(0.5))

1.0

In [25]:
- ((8 / 14) * log2(8 / 14)) - ((6 / 14) * log2(6 / 14))

0.9852281360342516

In [26]:
- ((2 / 6) * log2(2 / 6)) - ((4 / 6) * log2(4 / 6))

0.9182958340544896

In [27]:
((14 / 20) * 0.98) + ((6 / 20) * 0.91)

0.959

child node is more pure than parent node.

## Entropy Based on Class Split

In [28]:
- (0.5 * log2(0.5)) - (0.5 * log2(0.5))

1.0

In [30]:
- ((2 / 10) * log2(2 / 10)) - ((8 / 10) * log2(8 / 10))

0.7219280948873623

In [31]:
- ((8 / 10) * log2(8 / 10)) - ((2 / 10) * log2(2 / 10))

0.7219280948873623

In [32]:
(0.5 * 0.722) + (0.5 * 0.722)

0.722

| Performance | Class 
| 0.959       | 0.722

Information Gain 
Performance | 0.041
Class       | 0.278

Higher the information gain the higher the impurity

# Reduction Variance

Variance = (∑〖 [n_child  ∗ (X − μ) ^ 2] 〗) / n  

Where,
	X - Sample(0, 1)
	μ - Mean of samples
	n_child  - number of samples with X value(1 or 0)


In [34]:
((8 * 1) + (2 * 0)) / 10

0.8

In [41]:
((8 * ((1 - 0.8) ** 2)) + (2 * ((0 - 0.8) ** 2))) / 10

0.16

In [42]:
((8 * 0) + (2 * 1)) / 10

0.2

In [44]:
((2 * ((1 - 0.2) ** 2)) + (8 * ((0 - 0.2) ** 2))) / 10

0.16000000000000003

In [45]:
(((10 / 20) * 0.16) + ((10 / 20) * 0.16))

0.16

Performance | Class
0.238       | 0.16

smaller the variance the better the split. Thus class is better split